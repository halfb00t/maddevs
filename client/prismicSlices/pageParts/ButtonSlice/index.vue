<template>
  <div
    class="button-slice"
    :style="{ backgroundColor: sliceBackground }"
  >
    <div
      class="container"
      :class="`button-slice__content--align-${sliceAlignment}`"
      :data-aos="animation"
      data-testid="button-slice-container"
    >
      <LazyHydrate
        v-if="variation === 'default-slice'"
        when-visible
      >
        <UIButton
          :full-width="fullWidth"
          is-link
          :to="link.url"
          target="_self"
          data-testid="button-ui"
          :class="[
            'button-slice__button',
            'button-slice__button--normal',
          ]"
          :style="{ maxWidth }"
        >
          {{ text }}
        </UIButton>
      </LazyHydrate>
      <LazyHydrate
        v-else-if="variation === 'outlinedButton'"
        when-visible
      >
        <UIOutlinedButton
          :full-width="fullWidth"
          :color="color"
          is-link
          :to="link.url"
          target="_self"
          data-testid="button-outline"
          :class="[
            'button-slice__button',
            'button-slice__button--outlined',
          ]"
          :style="{ maxWidth }"
        >
          {{ text }}
        </UIOutlinedButton>
      </LazyHydrate>
      <LazyHydrate
        v-else-if="variation === 'linkButton'"
        when-visible
      >
        <UILinkButton
          :full-width="fullWidth"
          :color="color"
          is-link
          :to="link.url"
          target="_self"
          data-testid="button-link"
          :class="[
            'button-slice__button',
            'button-slice__button--link',
          ]"
          :style="{ maxWidth }"
        >
          {{ text }}
        </UILinkButton>
      </LazyHydrate>
    </div>
  </div>
</template>

<script>
import LazyHydrate from 'vue-lazy-hydration'
import animateOnScrollMixin from '@/mixins/animateOnScrollMixin'
import setSliceBackground from '@/helpers/setSliceBackground'

export default {
  name: 'ButtonSlice',
  components: {
    LazyHydrate,
    UIButton: () => import('@/components/shared/UIButton'),
    UIOutlinedButton: () => import('@/components/shared/UIOutlinedButton'),
    UILinkButton: () => import('@/components/shared/UILinkButton'),
  },

  mixins: [animateOnScrollMixin({
    offset: 200,
    delay: 50,
    anchorPlacement: 'top-center',
    duration: 1000,
    once: true,
  })],

  props: {
    slice: {
      type: Object,
      required: true,
      default() {
        return {}
      },
    },
  },

  data() {
    return {
      variation: this.slice?.variation,
      fullWidth: this.slice?.primary?.fullWidth,
      maxWidth: this.slice?.primary?.maxWidth,
      link: this.slice?.primary?.link,
      text: this.slice?.primary?.text,
      color: this.slice?.primary?.color,
      animation: this.slice?.primary?.animation,
      sliceBackground: setSliceBackground(this.slice?.primary?.background),
    }
  },

  computed: {
    sliceAlignment() {
      if (this.slice?.primary?.alignment === 'center') return 'center'
      if (this.slice?.primary?.alignment === 'right') return 'right'
      return 'left'
    },
  },
}
</script>

<style lang="scss" scoped>
.button-slice {
  display: block;
  margin-top: -1px;

  &__content {
    &--align {
      &-center {
        text-align: center;
      }
      &-right {
        text-align: right;
      }
    }
  }

  &__button {
    box-sizing: border-box;
    padding: 16px 24px;
    line-height: 20px;
    &--normal:hover {
      background-color: lighten($bgcolor--red, 10%) !important;
    }
  }
}
</style>
