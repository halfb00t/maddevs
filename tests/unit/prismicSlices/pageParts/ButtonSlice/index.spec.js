import { render, screen } from '@testing-library/vue'
import { shallowMount } from '@vue/test-utils'
import ButtonSlice from '@/prismicSlices/pageParts/ButtonSlice'

const backgrounds = {
  white: '#ffffff',
  black: '#111213',
  grey: '#f5f7f9',
}

const apiData = {
  background: 'black',
  animation: 'fade-up',
  link: {
    url: 'url',
  },
  alignment: 'center',
}

const getProps = params => ({
  slice: {
    primary: {
      ...params,
    },
  },
})

const stubs = ['UILinkButton', 'LazyHydrate', 'UIOutlinedButton', 'UIButton']

describe('Button slice', () => {
  it('should correctly render UIButton', () => {
    const props = getProps(apiData)
    props.slice.variation = 'default-slice'

    const { container } = render(ButtonSlice, {
      props,
      stubs,
    })

    expect(screen.queryByTestId('button-ui')).not.toBeNull()
    expect(screen.queryByTestId('button-outline')).toBeNull()
    expect(screen.queryByTestId('button-link')).toBeNull()
    expect(container).toMatchSnapshot()
  })

  it('should correctly render UIOutlinedButton', () => {
    const props = getProps(apiData)
    props.slice.variation = 'outlinedButton'

    const { container } = render(ButtonSlice, {
      props,
      stubs,
    })

    expect(screen.queryByTestId('button-ui')).toBeNull()
    expect(screen.queryByTestId('button-outline')).not.toBeNull()
    expect(screen.queryByTestId('button-link')).toBeNull()
    expect(container).toMatchSnapshot()
  })

  it('should correctly render UILinkButton', () => {
    const props = getProps(apiData)
    props.slice.variation = 'linkButton'

    const { container } = render(ButtonSlice, {
      props,
      stubs,
    })

    expect(screen.queryByTestId('button-ui')).toBeNull()
    expect(screen.queryByTestId('button-outline')).toBeNull()
    expect(screen.queryByTestId('button-link')).not.toBeNull()
    expect(container).toMatchSnapshot()
  })

  describe('Button slice background computed method', () => {
    it('should return background white', () => {
      const wrapper = shallowMount(ButtonSlice, {
        propsData: getProps({
          ...apiData,
          background: 'white',
        }),
      })

      expect(wrapper.vm.sliceBackground).toBe(backgrounds.white)
    })

    it('should return background grey', () => {
      const wrapper = shallowMount(ButtonSlice, {
        propsData: getProps({
          ...apiData,
          background: 'grey',
        }),
      })

      expect(wrapper.vm.sliceBackground).toBe(backgrounds.grey)
    })

    it('should return background black', () => {
      const wrapper = shallowMount(ButtonSlice, {
        propsData: getProps({
          ...apiData,
          background: 'black',
        }),
      })

      expect(wrapper.vm.sliceBackground).toBe(backgrounds.black)
    })

    it('should return null', () => {
      const wrapper = shallowMount(ButtonSlice, {
        propsData: getProps({
          ...apiData,
          background: 'unknown',
        }),
      })

      expect(wrapper.vm.sliceBackground).toBeNull()
    })
  })

  describe('Button slice data-aos animation attribute', () => {
    it('should container the data-aos to be fade-up', () => {
      render(ButtonSlice, {
        props: getProps(apiData),
      })

      expect(screen.getByTestId('button-slice-container').getAttribute('data-aos')).toBe(apiData.animation)
    })

    it('should container the data-aos to be null', () => {
      render(ButtonSlice, {
        props: getProps({}),
      })
      expect(screen.getByTestId('button-slice-container').getAttribute('data-aos')).toBeNull()
    })
  })

  describe('Button slice href attribute', () => {
    it('UIButton should render correctly url', () => {
      const wrapper = shallowMount(ButtonSlice, {
        propsData: getProps({
          ...apiData,
        }),
      })

      expect(wrapper.vm.$props.slice.primary.link.url).toBe(apiData.link.url)
    })

    it('UIButton should render null', () => {
      const wrapper = shallowMount(ButtonSlice, {
        propsData: getProps({}),
      })

      expect(wrapper.vm.$props.slice.primary.link).toBeUndefined()
    })
  })

  describe('Button slice alignment computed method', () => {
    it('should return alignment center', () => {
      const wrapper = shallowMount(ButtonSlice, {
        propsData: getProps({
          ...apiData,
        }),
      })

      expect(wrapper.vm.sliceAlignment).toBe(apiData.alignment)
    })

    it('should return alignment left', () => {
      const wrapper = shallowMount(ButtonSlice, {
        propsData: getProps({
          ...apiData,
          alignment: '',
        }),
      })

      expect(wrapper.vm.sliceAlignment).toBe('left')
    })

    it('should return alignment right', () => {
      const wrapper = shallowMount(ButtonSlice, {
        propsData: getProps({
          ...apiData,
          alignment: 'right',
        }),
      })

      expect(wrapper.vm.sliceAlignment).toBe('right')
    })
  })
})
