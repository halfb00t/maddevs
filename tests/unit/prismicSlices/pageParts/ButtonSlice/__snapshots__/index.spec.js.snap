// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button slice should correctly render UIButton 1`] = `
<div>
  <div
    class="button-slice"
    style="background-color: rgb(17, 18, 19);"
  >
    <div
      class="container button-slice__content--align-center aos-init aos-animate"
      data-aos="fade-up"
      data-testid="button-slice-container"
    >
      <lazyhydrate-stub
        idletimeout="2000"
        whenvisible="true"
      >
        <uibutton-stub
          class="button-slice__button button-slice__button--normal"
          data-testid="button-ui"
          is-link=""
          target="_self"
          to="url"
        >
          
        
      
        </uibutton-stub>
      </lazyhydrate-stub>
    </div>
  </div>
</div>
`;

exports[`Button slice should correctly render UILinkButton 1`] = `
<div>
  <div
    class="button-slice"
    style="background-color: rgb(17, 18, 19);"
  >
    <div
      class="container button-slice__content--align-center aos-init aos-animate"
      data-aos="fade-up"
      data-testid="button-slice-container"
    >
      <lazyhydrate-stub
        idletimeout="2000"
        whenvisible="true"
      >
        <uilinkbutton-stub
          class="button-slice__button button-slice__button--link"
          data-testid="button-link"
          is-link=""
          target="_self"
          to="url"
        >
          
        
      
        </uilinkbutton-stub>
      </lazyhydrate-stub>
    </div>
  </div>
</div>
`;

exports[`Button slice should correctly render UIOutlinedButton 1`] = `
<div>
  <div
    class="button-slice"
    style="background-color: rgb(17, 18, 19);"
  >
    <div
      class="container button-slice__content--align-center aos-init aos-animate"
      data-aos="fade-up"
      data-testid="button-slice-container"
    >
      <lazyhydrate-stub
        idletimeout="2000"
        whenvisible="true"
      >
        <uioutlinedbutton-stub
          class="button-slice__button button-slice__button--outlined"
          data-testid="button-outline"
          is-link=""
          target="_self"
          to="url"
        >
          
        
      
        </uioutlinedbutton-stub>
      </lazyhydrate-stub>
    </div>
  </div>
</div>
`;
